File: programmers/전력망을 둘로 나누기.py
"""
코딩테스트 연습 > 완전탐색
"""


from copy import deepcopy


def solution(n, wires):
    max_diff = 1000
    for i in range(len(wires)):
        new_wires = deepcopy(wires)
        new_wires.pop(i)
        new_wires = [p for pair in new_wires for p in (pair, pair[::-1])]
        visited = bfs([1], [], new_wires)

        diff = abs(n - 2 * len(visited))
        if max_diff > diff:
            max_diff = diff

    return max_diff


def bfs(queue, visited, wires):
    if len(queue) == 0:
        return visited

    q = queue.pop(0)
    visited.append(q)
    queue.extend(list(filter(lambda x: x not in visited, map(lambda x: x[1], filter(lambda x: x[0] == q, wires)))))
    return bfs(queue, visited, wires)


print(solution(9, [[1,3],[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]))
print(solution(4, [[1,2],[2,3],[3,4]]))
print(solution(7, [[1,2],[2,7],[3,7],[3,4],[4,5],[6,7]]))

File: programmers/전력망을 둘로 나누기.kt
import java.util.*
import kotlin.math.*

class Solution {
    fun solution(n: Int, wires: Array<IntArray>): Int {

        fun bfs(start: Int, wireList: MutableList<IntArray>): Int {
            var routes = mutableMapOf<Int, MutableSet<Int>>()
            wireList.forEach { wire ->
                if (routes[wire[0]] == null) {
                    routes[wire[0]] = mutableSetOf<Int>()
                }
                routes[wire[0]]!!.add(wire[1])

                if (routes[wire[1]] == null) {
                    routes[wire[1]] = mutableSetOf<Int>()
                }
                routes[wire[1]]!!.add(wire[0])
            }

            var visited = BooleanArray(n+1)
            var current = mutableListOf(start)

            var queue = ArrayDeque<Int>()
            queue.addLast(start)
            while (queue.size > 0) {
                var now = queue.pollFirst()
                if (!visited[now]) {
                    visited[now] = true
                    routes[now]?.let{
                        queue.addAll(it.toList())
                    }
                }
            }

            return visited.filter{it==true}.size
        }

        var wireList = wires.toMutableList()
        var networkDiffList = mutableListOf<Int>()
        repeat(wires.size) {
            var targetWire = wireList.removeAt(0)

            var networkA = bfs(1, wireList)
            var networkB = n - networkA
            networkDiffList.add(abs(networkA - networkB))

            wireList.add(targetWire)
        }

        return networkDiffList.sorted().first()
    }
}
